macros/src/bitflags.rs

```diff
@@ -0,0 +1,153 @@
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{quote, ToTokens};
+use syn::{
+    parse::{Parse, ParseStream},
+    parse_macro_input,
+    punctuated::Punctuated,
+    token, Attribute, Expr, Ident, Token, Type, Visibility,
+};
+
+use crate::{mksym, symbol::Symbol};
+
+pub(super) fn expand(ts: TokenStream) -> TokenStream {
+    let ts2 = proc_macro2::TokenStream::from(ts.clone());
+    let input: Input = parse_macro_input!(ts);
+
+    let codegen = Codegen::new(&input);
+
+    // Encode package and disambiguator to provide the decoder with all info it needs (even if
+    // technically redundant, since it's also stored in the symbol we create).
+    let format_str = format!(
+        "{{={}:__internal_bitflags_{}@{}@{}}}",
+        input.ty.to_token_stream(),
+        input.ident,
+        crate::symbol::package(),
+        crate::symbol::disambiguator(),
+    );
+    let sym = mksym(&format_str, "bitflags", false);
+
+    let ident = &input.ident;
+    let ty = input.ty;
+    let flags = codegen.flags.iter().map(|f| &f.def);
+    let result = quote! {
+        const _: () = {
+            fn assert<T: defmt::export::UnsignedInt>() {}
+            assert::<#ty>;
+
+            #(#flags)*
+        };
+
+        defmt::export::bitflags! {
+            #ts2
+        }
+
+        impl defmt::Format for #ident {
+            fn format(&self, f: defmt::Formatter) {
+                unreachable!()
+            }
+            fn _format_tag() -> defmt::Str {
+                #sym
+            }
+            fn _format_data(&self) {
+                // There's a method available for every supported bitflags type.
+                defmt::export::#ty(&self.bits());
+            }
+        }
+    };
+    result.into()
+}
+
+struct FlagStatic {
+    def: TokenStream2,
+}
+
+struct Codegen {
+    flags: Vec<FlagStatic>,
+}
+
+impl Codegen {
+    fn new(input: &Input) -> Self {
+        let flags = input
+            .flags
+            .iter()
+            .map(|flag| {
+                let name = &flag.ident;
+                let value = &flag.value;
+                let repr_ty = &input.ty;
+
+                let sym = Symbol::new(
+                    "bitflags_value",
+                    &format!("{}::{}", input.ident, flag.ident),
+                )
+                .mangle();
+
+                let def = quote! {
+                    #[cfg_attr(target_os = "macos", link_section = ".defmt,end")]
```

Comment:
(the difference between these two cfg attributes is quite hard to spot: `.` vs `,`)